proxy/coco/src/request/waiting_room/stream.rs

```diff
@@ -0,0 +1,264 @@
+//! Streaming interfaces for requests in the [`WaitingRoom`].
+
+use std::{
+    future::Future as _,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+use futures::Stream;
+use tokio::{
+    sync::RwLock,
+    time::{delay_for, Delay},
+};
+
+use librad::uri::{RadUrl, RadUrn};
+
+use crate::request::waiting_room::WaitingRoom;
+
+/// How much a stream should delay before polling the waiting room again.
+const REQUEST_DELAY: Duration = Duration::from_millis(20);
+
+/// A stream of queryable requests. See [`Queries::new`] for more information.
+pub struct Queries {
+    /// How long the stream should delay before each poll of the waiting room.
+    delay: Delay,
+    /// The waiting room that will be polled for retrieving the requests.
+    waiting_room: Arc<RwLock<WaitingRoom<Instant, Duration>>>,
+}
+
+impl Queries {
+    /// Create a new `Queries` stream.
+    ///
+    /// This type implements [`Stream`] and is expected to be consumed through that interface.
+    ///
+    /// The stream will return the next request that is available for querying. This is driven by
+    /// the supplied `waiting_room` -- as more requests transtion through it. If no such request
+    /// exists then the stream does nothing and waits until the next poll. We note that the stream
+    /// is infinite with this respect and will never return `None`.
+    #[must_use]
+    pub fn new(waiting_room: Arc<RwLock<WaitingRoom<Instant, Duration>>>) -> Self {
+        Self {
+            delay: delay_for(REQUEST_DELAY),
+            waiting_room,
+        }
+    }
+
+    /// Reset the delay for the stream.
+    fn reset_delay(&mut self) {
+        self.delay = delay_for(REQUEST_DELAY)
+    }
+}
+
+impl Stream for Queries {
+    type Item = RadUrn;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let waiting_room = self.waiting_room.clone();
+        let waiting_room = futures::executor::block_on(waiting_room.read());
```

Comment:
Uh-oh, blocking is verboten. You need to store the untypeable future returned from `RwLock::read()` in `self`, and poll that. If it returns you the guard, proceed, else return `Poll::Pending`.

At least, I don't know a better way to desugar `async/await` here.

Reply:
Thanks, that might actually be the cause of our previous bug :thinking: Created a follow up issue: https://github.com/radicle-dev/radicle-upstream/issues/981